### DISCUSSION TITLE
[Java][MS Edge] Change default download location

### DISCUSSION THREAD

**Discussion by email22lead** [selenide/selenide#2345] [Q&A]
Hello Everyone,

I have a problem I can't resolve. Using Selenide I'm not able to change the Edge browser's download location. Tried msedge-selenium-tools-java way to use specific Profile in Edge (code crashes), tried setting the path through the Configuration (seems to be not working at all). I'm running out of ideas here. Could you please help? Not sure how to proceed with proxy download, couldn't find any example code.

Edge version 114, Selenide 6.12

Best regards

**Comment by asolntsev** [comment]
Hi @email22lead .
The answer is simple:
1. Edge doesn't support setting custom downloads folder, BUT
2. You **don't need to set custom downloads folder**. Downloading files works fine in Selenide even without setting custom folder. Just use method `File report = $("#report").download();`.

**Comment by email22lead** [comment]
> Hi @email22lead . The answer is simple:
> 
> 1. Edge doesn't support setting custom downloads folder, BUT

Oh, good to know then.

> 2. You **don't need to set custom downloads folder**. Downloading files works fine in Selenide even without setting custom folder. Just use method `File report = $("#report").download();`.

Yeah, unfortunately my element is a //a link element with text attribute. Once clicked it starts downloading a file. There's no href attribute. Is there any code example how to set a proxy to download a file to the location which I want?

**Reply by asolntsev** [reply]
@email22lead  It's always needed to use proxy.
The simplest way to download files without `href` attribute is `FOLDER`:

https://selenide.org/2020/07/08/selenide-5.13.0/#new-file-download-mode-folder

Just set `Configuration.fileDownload = FileDownloadMode.FOLDER;` and call the old good `$.download();`.

**Comment by email22lead** [comment]
Hi @asolntsev,

thanks for a suggestion.
Unfortunately, the file is still being downloaded to user's Downloads folder.

Code snippet:

```
try {
            Configuration.downloadsFolder = "build\\downloads";
            File downloadedFile = exportBtn.download();
            LOGGER.info("Downloaded file name: " + downloadedFile.getName());
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
```

download() methods works fine, it clicks on the element, the file is downloaded but not to the expected location, so the FileNotFoundException is thrown. Is there any way to overcome this on MS Edge?

**Reply by asolntsev** [reply]
I think, you don't need to change the folder. Just remove line `Configuration.downloadsFolder = "build\\downloads";`.

P.S. If you really need to change `Configuration.downloadsFolder`, you should do it BEFORE opening the browser. But believe me, you don't need it.

**Reply by email22lead** [reply]
Used it before opening the browser, no luck.
Removed the `Configuration.downloadsFolder = "build\\downloads";`, no luck. Seems like the method is not working fine on Edge. It still downloads the file to user's Download folder. In build folder, the downloads folder is not even created (I remove it every time before downloading a file). I guess it should be created by default if the file is meant to be downloaded to this location.

Anyway, no luck on Edge still. Am I missing something?

**Reply by asolntsev** [reply]
@email22lead  Sorry. In my previous comment, I was mistaken ("Edge doesn't support setting custom downloads folder"). Actually, Edge allows custom downloads folder. 

But it doesn't really matter. Just **DON'T set custom downloads folder**. Don't set `Configuration.downloadsFolder = ...`

Just call method `$.download()`. If you think it's still not working, please share the whole Selenide log.
Among others, it should contain something like this:

```
09:27:04:620 [main] INFO WebDriverManager - Exporting webdriver.edge.driver as /Users/andrei/.cache/selenium/msedgedriver/mac-arm64/114.0.1823.55/msedgedriver

09:27:11:647 [main] DEBUG AbstractChromiumDriverFactory - Using chromium preferences: {credentials_enable_service=false, safebrowsing.enabled=true, plugins.always_open_pdf_externally=true, profile.default_content_setting_values.automatic_downloads=1, download.default_directory=/Users/andrei/projects/selenide/statics/build/downloads/1687415214020_3466_1}

09:27:13:611 [main] DEBUG DownloadFile - Download file: method: FOLDER, timeout: 5000 ms, filter: with extension "txt"
09:27:22:570 [main] DEBUG BrowserDownloadsFolder - Going to clean folder /Users/andrei/projects/selenide/statics/build/downloads/1687415214020_3466_1 - found files: []
09:27:25:857 [main] DEBUG BrowserDownloadsFolder - After clean folder /Users/andrei/projects/selenide/statics/build/downloads/1687415214020_3466_1: []
09:27:26:352 [main] DEBUG DownloadFileToFolder - Waiting for files in /Users/andrei/projects/selenide/statics/build/downloads/1687415214020_3466_1...
09:27:26:368 [main] DEBUG DownloadFileToFolder - Matching files not found: [], all new files: .com.microsoft.edgemac.IM2kfs (modified 0 ms ago) (1 file at 1687415246367), all files: [.com.microsoft.edgemac.IM2kfs]
09:27:26:646 [main] DEBUG DownloadFileToFolder - Matching files found: [large_file.txt (modified 148 ms ago)], all new files: large_file.txt (modified 147 ms ago) (1 file at 1687415246646), all files: [large_file.txt]
09:27:26:647 [main] DEBUG DownloadFileToFolder - No [crdownload, tmp] files found, conclude download is completed (filter: with extension "txt")
09:27:26:648 [main] INFO DownloadFileToFolder - Downloaded files in /Users/andrei/projects/selenide/statics/build/downloads/1687415214020_3466_1: large_file.txt (modified 150 ms ago) (1 file at 1687415246648)
09:27:26:648 [main] DEBUG DownloadFileToFolder - All downloaded files: [large_file.txt]

```
